src/Canonical.php

```diff
@@ -44,27 +44,39 @@ class Canonical
     /** @var RouterInterface */
     private $router;
 
+    /** @var RequestStack */
+    private $requestStack;
+
     public function __construct(Config $config, UrlGeneratorInterface $urlGenerator, RequestStack $requestStack, RouterInterface $router, string $defaultLocale)
     {
         $this->config = $config;
         $this->urlGenerator = $urlGenerator;
-        $this->request = $requestStack->getCurrentRequest() ?? Request::createFromGlobals();
         $this->defaultLocale = $defaultLocale;
         $this->router = $router;
+        $this->requestStack = $requestStack;
+    }
+
+    public function getRequest(): Request
+    {
+        // always try to use current request
+        if ($this->request === null) {
+            $this->request = $this->requestStack->getCurrentRequest() ?? Request::createFromGlobals();
+            $this->init();
+        }
 
-        $this->init();
+        return $this->request;
     }
 
     public function init(): void
     {
         // Ensure in request cycle (even for override).
-        if ($this->request === null || $this->request->getHost() === '') {
+        if ($this->getRequest() === null || $this->getRequest()->getHost() === '') {
```

Comment:
Fix PHPStan's remark: 

> Strict comparison using === between Symfony\Component\HttpFoundation\Request and null         will always evaluate to false.

```suggestion
        if ($this->getRequest()->getHost() === '') {
```

Reply:
Agreed with possible infinite loop.
I aimed to keep `init()` for backwards compatibility.
But maybe we should simply ditch it.

Reply:
See latest changes.
I added a request setter.
Please re-run checks.

My tests passed.